File: nezal-ar/public/base-defender/js/game/isometric/map.js
var Map={
	scene : null,
	tileWidth:64,
	tileHeight:31,
	viewWidth:760,
	viewHeight:550,
	mapWidth : 1000,
	mapHeight : 1000,	
	hoverX :0,
	hoverY :0,
	x:380,y:275,
	speed:2,
	origin:[0,0],
	movementSpeed : 8,
	rowEvenDirections : {},
	rowOddDirections : {},
	objects : [],
	inViewObjects : [],
	init : function (scene){
		this.scene = scene;
		var rawMap = scene.rawMap;
		this.grid=[];
			for(var i=0;i<rawMap.length;i++){
					for(var j=0;j<rawMap[0].length;j++){
						if(!Map.grid[i])Map.grid[i]=[];
						Map.grid[i][j] = new Node(i,j);
						Map.grid[i][j].terrainType = rawMap[i][j];
						if(Map.grid[i][j].terrainType != 0){
							scene.renderDisplayUnit(Map.value(i, j), rawMap[i][j]);
					}
				}
			}			
			Map.navigation = new Navigation(Map);
			Map.mapMirror = new MapMirror();
			Map.initDirections()
			Map.tileIsoLength = Math.sqrt(Math.pow(Map.tileWidth/2,2)+Math.pow(Map.tileHeight/2,2))/2
			Map.tileAngle = Map.getTileAngle()
	},
	
	initializeMapSize : function(){
		this.containerDiv = $("gameSceneContainer")
		this.div = $('gameCanvas');
		this.div.style.backgroundImage = "url(images/game_elements/background.png)"
		Map.containerDiv.style.width = Map.viewWidth + "px";
		Map.containerDiv.style.height = Map.viewHeight + "px";
		var backgroundImg = Loader.images.game_elements['background.png'];
		Map.div.style.width  =  backgroundImg.width + "px";
		Map.mapWidth = backgroundImg.width;
		Map.div.style.height =  backgroundImg.height + "px";
		Map.mapHeight = backgroundImg.height;
		Map.centerMap(1)
	},
	centerMap : function(zoomFactor){
		Map.x = 0
		Map.y=0
		Map.move(Map.mapWidth/2-Map.viewWidth/2,Map.mapHeight/2-Map.viewHeight/2)
	},
	
	initDirections : function(){
		Map.rowEvenDirections[Map.N]=[-2,0]
		Map.rowEvenDirections[Map.S]=[2,0]
		Map.rowEvenDirections[Map.E]=[0,1]
		Map.rowEvenDirections[Map.W]=[0,-1]
		Map.rowEvenDirections[Map.NE]= [-1,0]
		Map.rowEvenDirections[Map.NW]=[-1,-1] 
		Map.rowEvenDirections[Map.SE]=[1,0]
		Map.rowEvenDirections[Map.SW]=[1,-1]
		Map.rowOddDirections[Map.N]=[-2,0]
		Map.rowOddDirections[Map.S]=[2,0]
		Map.rowOddDirections[Map.E]=[0,1]
		Map.rowOddDirections[Map.W]=[0,-1]
		Map.rowOddDirections[Map.NE]=[-1,1]
		Map.rowOddDirections[Map.NW]=[-1,0]
		Map.rowOddDirections[Map.SE]=[1,1]
		Map.rowOddDirections[Map.SW]=[1,0]
	},
  getGeneralDirectoin : function(x1,y1,x2,y2) {
    var dir = 0
    var slope = Math.round((y1-y2)/(x1-x2))
    if(slope == "Infinity")
      dir = Map.N;
    else if (slope == "-Infinity" )
      dir = Map.S;
    else if(slope == 0 && x1 > x2)
      dir = Map.W
    else if(slope == 0 && x1 < x2)
      dir = Map.E
    else if(slope == -1 && x1 < x2)
      dir = Map.NE
    else if(slope < -1 && x1 < x2)
      dir = Map.N
    else if(slope == -1 && x1 > x2)
      dir = Map.SW
    else if(slope < -1 && x1 > x2)
      dir = Map.S
    else if(slope == 1 && x1 > x2)
      dir = Map.NW
    else if(slope > 1 && x1 > x2)
      dir = Map.N
    else if(slope == 1 && x1 < x2)
      dir = Map.SE
    else if(slope > 1 && x1 < x2)
      dir = Map.S
    return dir;
  },
	getTileAngle : function(){
		var a = Math.sqrt(Math.pow(Map.tileWidth/2,2)+Math.pow(Map.tileHeight/2,2))/2
		var theta = (180-Util.radToDeg(Math.acos((Map.tileWidth*Map.tileWidth/4-2*a*a)/(2*a*a))))/2
		return theta
	},
	getDirection : function(x1,y1,x2,y2){
		var tile1 = Map.tileValue(x1,y1)
		var tile2 = Map.tileValue(x2,y2)
		i = tile1[0]; j = tile1[1]
		i2 = tile2[0]; j2 = tile2[1]
		var x = -1,y=-1;
		var rowEvenDirections = [[Map.N,i-2,j],[Map.S,i+2,j],[Map.E,i,j+1],[Map.W,i,j-1],[Map.NE,i-1,j],[Map.NW,i-1,j-1],[Map.SE,i+1,j],[Map.SW,i+1,j-1]]
		var rowOddDirections = [[Map.N,i-2,j],[Map.S,i+2,j],[Map.E,i,j+1],[Map.W,i,j-1],[Map.NE,i-1,j+1],[Map.NW,i-1,j],[Map.SE,i+1,j+1],[Map.SW,i+1,j]]
		if(i%2 == 0){
			for(var k=0;k<rowEvenDirections.length;k++){
				if(i2==rowEvenDirections[k][1]&&j2==rowEvenDirections[k][2]) return rowEvenDirections[k][0]
			}
		}else{
			for(var k=0;k<rowOddDirections.length;k++){
				if(i2==rowOddDirections[k][1]&&j2==rowOddDirections[k][2]) return rowOddDirections[k][0]
			}
		}
		return 0
	},
	getNeighbor : function(i,j,direction){
		var ret = null
		if(i%2==0) ret = Map.rowEvenDirections[direction]
		else ret = Map.rowOddDirections[direction]
		if(Map.grid[ret[0]+i]&&Map.grid[ret[0]+i][ret[1]+j]) return [ret[0]+i, ret[1]+j]
		return null
	},
	neighbors : function(node){
		var i=node.x
		var j=node.y
		var retList = []
		for (var k=0;k<8;k++){
			var neighbor = Map.getNeighbor(i,j,k)
			if(neighbor && (Map.grid[neighbor[0]][neighbor[1]].value ==0 || Map.grid[neighbor[0]][neighbor[1]].target)  &&Map.grid[neighbor[0]][neighbor[1]].terrainType==0 ) retList.push(Map.grid[neighbor[0]][neighbor[1]])
		}
		return retList
	},
	getAngleBetweenTiles : function(x1,y1,x2,y2){
		var tile1 = Map.tileValue(x1,y1)
		var tile2 = Map.tileValue(x2,y2)
		var dir = (x1%2==0)? -1 : 1
		var angle1 = Map.tileAngle
		var angle2 = 180 - Map.tileAngle
		var angle3 = 180 + Map.tileAngle
		var angle4 = 360 - Map.tileAngle
		var itr = []
		if(dir == 1) itr = [[-1,0,1],[1,0,3],[-1,dir,5],[1,dir,7],[-2,0,2],[2,0,6],[0,1,0],[0,-1,4]]
		else itr = [[-1,0,5],[1,0,7],[-1,dir,1],[1,dir,3],[-2,0,2],[2,0,6],[0,1,0],[0,-1,4]]
		for(var k=0;k< itr.length; k++){
				var l = tile1[0]+itr[k][0]
				var m = tile1[1]+itr[k][1]
				if(l==tile2[0]&&m==tile2[1]){
					return itr[k][2]
				}
		}
		return 0
	},
	
	//takes x,y values and map them to the map tiles
	tileValue : function(x,y){
		var halfX = Math.floor(x*2/Map.tileWidth)
		var halfY = Math.floor(y*2/Map.tileHeight)
		var halfW = Map.tileWidth/2
		var halfH = Map.tileHeight/2
		var up = false
		if(halfX%2==halfY%2){
			if(x-halfX*halfW+y-halfY*halfH < (halfX+1)*halfW-x+(halfY+1)*halfH-y){
				up = true
			}
		}else{
			if((halfX+1)*halfW-x+y-halfY*halfH <x-halfX*halfW +(halfY+1)*halfH-y){
				up = true
			}
		}
		var gridX=0
		var gridY=0
		if(halfX%2==0&&halfY%2==0&&!up ||halfX%2!=0&&halfY%2==0&&!up || halfX%2!=0&&halfY%2!=0 &&up || halfX%2==0&&halfY%2!=0 &&up)
		{
			//green area
			gridY = 2*Math.floor(halfY/2)+1
			gridX = Math.floor(halfX/2)
		}
		else{
			//brown area
			gridY = halfY + halfY%2
			gridX = Math.floor((halfX+1)/2)
		}
		return [gridY,gridX]
		
	},
	
	//takes tile values and map them to x,y values
	value : function(i,j){
		var y=Math.max(Math.round((i-1)*Map.tileHeight/2),0) 
		var	x=-((i+1)%2)*Map.tileWidth/2+j*Map.tileWidth+Map.tileWidth/2;
		return [x,y]
	},
	
	drawMap : function(){
		var img = new Image
		var img2 = new Image
		img.src="images/tile_dirt.png"
		img.onload = function(){
			img2.src="images/tile_grass.png"
			img2.onload = function(){
				//	Game.runGame()
			}
		}
		this.img = img
		this.img2 = img2
	},
	
	doDrawMap : function(){
		var origin = this.origin
		if(origin[0]>0)origin[0]--
		if(origin[1]>0)origin[1]--
		var end = Map.tileValue(Math.min(this.x+this.viewWidth,this.mapWidth),Math.min(this.y+this.viewHeight,this.mapHeight))
		for(i=0;10;i++){
			for(j=0;j<10;j++){
				var values = Map.value(i+origin[0],j+origin[1])
				var x=values[0]-this.x
				var y=values[1]-this.y
				if ((x<Map.viewWidth) && (y<Map.viewHeight) &&
					(x+Map.tileWidth>0) && (y+Map.tileHeight>0))
					{
						var img = null
						if(((origin[0]+i)%2 == 0)){
							img = this.img	
						}
						else{
							img = this.img2
						}
						img.style.position = "absolute"
						img.style.left = x +"px"
						img.style.top = y +"px"
						Map.div.appendChild(img.clone())
					}
			}
		}
		this.inViewObjects = []
		for(var k=0 ;k < this.objects.length ;k++){
			var obj=this.objects[k]
			if(obj.owner.x -this.x+Math.round(obj.imgWidth/2) >0 && obj.owner.x -this.x-Math.round(obj.imgWidth/2) < this.viewWidth 
			&& obj.y-this.y-Math.round(obj.imgHeight/2) < this.viewHeight && obj.y-this.y+Math.round(obj.imgHeight/2) >0)
			this.inViewObjects.pushInPlaceDescending(obj,'y')
		}

		var end = Map.tileValue(Math.min(this.x+this.viewWidth,this.mapWidth),Math.min(this.y+this.viewHeight,this.mapHeight))		
	},
	
	// gets the border tile for x,y values
	borderTileValues : function(x,y){
		var tileValues = Map.tileValue(x,y)
		return Map.value(tileValues[0],tileValues[1])		
	},

	addObjectToGrid : function(obj){
		var noOfRows = Math.ceil(obj.xdim/Map.tileIsoLength)
		var noOfColumns = Math.ceil(obj.ydim/Map.tileIsoLength)
		var topX = obj.owner.coords.x 
		var topY = obj.owner.coords.y -Map.tileHeight/2
		var originTile = Map.tileValue(topX,topY)
		var mapTiles = []
		for(var i=0;i<noOfRows+1;i++){
			if(!originTile) continue;
			originTile = Map.getNeighbor(originTile[0],originTile[1],Map.SW)
			if(!originTile) continue;
			var loopingTile = [];loopingTile[0] = originTile[0]; loopingTile[1] = originTile[1]
			for(var j=0;j<noOfColumns+1;j++){
				Map.grid[loopingTile[0]][loopingTile[1]].value = obj
				mapTiles.push(Map.grid[loopingTile[0]][loopingTile[1]])
				loopingTile = Map.getNeighbor(loopingTile[0],loopingTile[1],Map.SE)
			}
		}
		obj.mapTiles = mapTiles
	},
	
	addCreep : function(x,y){
		var creep = new Tank()
		creep.x = this.x+x
		creep.y = this.y+y
		Map.addElement(creep)
	},
	
	getRealCoords : function(x,y){
		return {"x":x+Map.x,"y":y+Map.y}
	},
	
	validateLocation : function(obj){
		var collide = false
		for(var i=0;i<this.objects.length;i++){
			if(obj.collides(this.objects[i])){
				collide = true
				break
			}			
		}
		return !collide
	},
	
	addElement : function(obj){
		var valid = this.validateLocation(obj)
		if(valid) {
	  	this.objects.push(obj)
	  	this.addObjectToGrid(obj)
	  }
		return valid
	},
	
	lookupLocation : function(x,y){
		var mapCoords = Map.tileValue(x,y)
		if(Map.grid[mapCoords[0]][mapCoords[1]].value)	return Map.grid[mapCoords[0]][mapCoords[1]].value.owner
		return null
	},
	clear :function(){
		if(this.objects)this.objects = []
		if(this.div)this.div.innerHTML = ""
	},
	occupied : function(x,y){
		var mapTiles = Map.tileValue(x,y)
		if(Map.grid[mapTiles[0]][mapTiles[1]].value)return true
		return false 
	},
	
	repeatMovement : function(xDirection,yDirection){
		var self = this;
		Map.periodicalMovementEvent = setInterval(function(){self.move(xDirection, yDirection)}, this.movementSpeed);
	},

	move : function (dx,dy){
		if(this.x+dx<0)
		{
			this.x=0;
		}
		else if (this.x+dx>this.mapWidth-this.viewWidth)
		{
			this.x = this.mapWidth - this.viewWidth
		}
		else
		{
			this.x+=dx;
			
		}
		if(this.y+dy<0)
		{
			this.y=0;
		}
		else if (this.y+dy>this.mapHeight-this.viewHeight)
		{
			this.y = this.mapHeight-this.viewHeight
		}
		else
		{
			this.y+=dy
		}
		this.div.style.left = -this.x +"px"
		this.div.style.top = -this.y +"px"
		this.origin = Map.tileValue(this.x,this.y)
		
		
	},
	
	stopMovement : function(){
		clearInterval(Map.periodicalMovementEvent);
	},
	
	hovered : function(x,y){
	},
	
	moveObject : function(object,x,y,callback,ignorePlace){
		var astar = new Astar()
		var srcTiles = Map.tileValue(object.coords.x,object.coords.y)
		var destTiles = Map.tileValue(x,y)
		Map.grid[destTiles[0]][destTiles[1]].target = true
	if(Map.grid[destTiles[0]][destTiles[1]].value!=0 && !ignorePlace) destTiles = Map.getNearestEmptyTile(destTiles)
		var path = astar.getOptimalPath(Map,Map.grid[srcTiles[0]][srcTiles[1]],Map.grid[destTiles[0]][destTiles[1]])
		for(var i=0;i<Math.ceil(this.mapHeight*2/this.tileHeight)+1;i++){
				for(var j=0;j<Math.ceil(this.mapWidth/this.tileWidth)+1;j++){
					this.grid[i][j].g=this.grid[i][j].h=this.grid[i][j].f=this.grid[i][j].visited=
					this.grid[i][j].closed = 0
					this.grid[i][j].parent = null
				}
			}
		Map.grid[destTiles[0]][destTiles[1]].target = false	
		if(path){
			object.moving = false;
			object.targetPoint = {x:x,y:y}
			object.movingPath = path;
			if(callback)object.movementFinishCallback = callback;
			return path 
		}
		
	},
	
	getNearestEmptyTile : function(tile){
		var openList = []
		openList.push(tile)
		var found = false
		while(!found){
			var tile = openList[0]
		 	for (var k=0;k<8;k++){
				var neighbor = Map.getNeighbor(tile[0],tile[1],k)
				if(neighbor && Map.grid[neighbor[0]][neighbor[1]].value ==0 &&Map.grid[neighbor[0]][neighbor[1]].terrainType==0 )return neighbor
				else if(neighbor) openList.push(neighbor)
			}
			openList.splice(0,1)			
		}	
	},
	
	registerListeners : function(div,owner){
		div.observe(game.mouseClickEvent,function(){
			if(!game.buildingMode.isOn && owner.working){
				owner.defaultAction();
			}
		})
		div.observe(game.mouseStartEvent,function(event){
			if (event.button != 2) {
			  	if (event.preventDefault) {
			  		event.preventDefault();
			  	}
	  		}
		})
    
    var mousemoveCallback = function(mouse){
      var x = mouse.pointerX() || mouse.touches[0].pageX;
      var y = mouse.pointerY() || mouse.touches[0].pageY;
      owner.sprites.mouseover.shiftX = x - ( owner.coords.x - Math.round(owner.imgWidth/2) - Map.x ) + 10;
      owner.sprites.mouseover.shiftY = y - ( owner.coords.y -Math.round(owner.imgHeight/2) - Map.y );
      owner.sprites.mouseover.render();
      owner.sprites.mouseover.show();
    }
    
    div.observe(game.mouseMoveEvent, mousemoveCallback);
    
		div.observe('mouseover',function(){
	  	if (owner.state != owner.states.NOT_PLACED) {
	  		if (owner.state != owner.states.UNDER_CONSTRUCTION)owner.sprites.outline.show();
  			if (owner.state == owner.states.NORMAL)owner.sprites.info.show();
	  		if(owner.sprites.text)owner.sprites.text.show()
        if(owner.state == owner.states.NORMAL && owner.sprites.moreContainer) 
          owner.sprites.moreContainer.show();
	  	}
		})
		div.observe('mouseout',function(){
			owner.sprites.outline.hide();
		    owner.sprites.mouseover.hide();
      		owner.sprites.info.hide();
			if(owner.sprites.text)owner.sprites.text.hide()
      if(owner.state == owner.states.NORMAL && owner.sprites.moreContainer) 
        owner.sprites.moreContainer.hide();
		})
	},

	registerSpecialListeners : function(div,owner, clickCall){
		div.observe(game.mouseClickEvent,function(){
			console.log('here')
			if(!game.buildingMode.isOn && owner.working){
				console.log('here')
				owner[clickCall]();
			}
		})
		div.observe(game.mouseStartEvent,function(event){
			if (event.button != 2) {
			  	if (event.preventDefault) {
			  		event.preventDefault();
			  	}
	  		}
		})
    
		div.observe('mouseover',function(){
	  	if (owner.state != owner.states.NOT_PLACED) {
	  		if (owner.state != owner.states.UNDER_CONSTRUCTION)owner.sprites.outline.show();
  			if (owner.state == owner.states.NORMAL)owner.sprites.info.show();
	  		if(owner.sprites.text)owner.sprites.text.show()
        if(owner.state == owner.states.NORMAL && owner.sprites.moreContainer) 
          owner.sprites.moreContainer.show();
	  	}
		})
		div.observe('mouseout',function(){
			owner.sprites.outline.hide();
		    owner.sprites.mouseover.hide();
      		owner.sprites.info.hide();
			if(owner.sprites.text)owner.sprites.text.hide()
      if(owner.state == owner.states.NORMAL && owner.sprites.moreContainer) 
        owner.sprites.moreContainer.hide();
		})
	},
//	checkBusyTiles : function(){
//		var canvas = document.createElement('canvas')
//		canvas.setAttribute('width',Map.mapWidth)
//		canvas.setAttribute('height',Map.map)
//	}
	E:0, NE:1, N:2, NW:3, W:4, SW:5, S:6, SE:7 
}

File: nezal-ar/app/models/games/base_defender/isometric/map.rb
module BD
  class Map
      @@E=0
      @@NE=1
      @@N=2
      @@NW=3
      @@W=4
      @@SW=5
      @@S=6
      @@SE=7
      ['E', 'NE', 'N', 'NW', 'W', 'SW', 'S', 'SE'].each do |attr|
  class_eval <<EOF
    def self.#{attr}
      @@#{attr}
    end
    
    def self.#{attr}=(new_value)
      @@#{attr} = new_value
    end
EOF
    end
    def initialize raw_map
      initialize_variables
      no_of_rows = raw_map.length
      no_of_columns = raw_map[0].length
      0.upto(no_of_rows-1) do |i|
        @grid[i]= []
        0.upto(no_of_columns-1) do |j|
          @grid[i][j] = Node.new(i,j)
           @grid[i][j].terrain_type = raw_map[i][j];
        end
      end
      init_directions()
      @tile_iso_length = Math.sqrt((@tile_width/2)**2+(@tile_height/2)**2)/2
      @tile_angle = get_tile_angle()
    end
    
    def initialize_variables
      @grid = []
      @tile_width=64.0
      @tile_height=31.0
      @view_width=760.0
      @view_height=550.0
      @map_width = 1000.0
      @map_height = 1000.0 
      @x=380.0,@y=275.0
      @speed=2.0
      @origin=[0.0,0.0]
      @row_even_directions = {}
      @row_odd_directions = {}
      @objects = []
      @zone_points = [{'x'=>1495,'y'=>551},{'x'=>765,'y'=>914},{'x'=>25,'y'=>550},{'x'=>771,'y'=>157}]
    end
    attr_accessor :grid, :tile_width, :tile_height, :view_width, :view_height, :map_width, :map_height, :x, :y, :speed,
      :origin, :row_even_directions, :row_odd_directions, :objects, :tile_angle
    def init_directions
     @row_even_directions[@@N]=[-2,0]
     @row_even_directions[@@S]=[2,0]
     @row_even_directions[@@E]=[0,1]
     @row_even_directions[@@W]=[0,-1]
     @row_even_directions[@@NE]= [-1,0]
     @row_even_directions[@@NW]=[-1,-1] 
     @row_even_directions[@@SE]=[1,0]
     @row_even_directions[@@SW]=[1,-1]
     @row_odd_directions[@@N]=[-2,0]
     @row_odd_directions[@@S]=[2,0]
     @row_odd_directions[@@E]=[0,1]
     @row_odd_directions[@@W]=[0,-1]
     @row_odd_directions[@@NE]=[-1,1]
     @row_odd_directions[@@NW]=[-1,0]
     @row_odd_directions[@@SE]=[1,1]
     @row_odd_directions[@@SW]=[1,0]
   end
    
   def get_tile_angle
    a = Math.sqrt((@tile_width/2)**2+(@tile_height/2)**2)/2
    theta = (180-Util.rad_to_deg(Math.acos((@tile_width*@tile_width/4-2*a*a)/(2*a*a))))/2
    theta
  end

  def self.get_general_direction x1,y1,x2,y2
    dir = 0
    slope = 0
    begin 
      slope = ((y1-y2).to_f/(x1-x2)).round
      if(slope == 0 && x1 > x2)
        dir = @@W
      elsif(slope == 0 && x1 < x2)
        dir = @@E
      elsif(slope == -1 && x1 < x2)
        dir = @@NE
      elsif(slope == -1 && x1 > x2)
        dir = @@SW
      elsif(slope == 1 && x1 > x2)
        dir = @@NW
      elsif(slope == 1 && x1 < x2)
        dir = @@SE
      elsif(slope < -1 && x1 < x2)
        dir = @@N
      elsif(slope < -1 && x1 > x2)
        dir = @@S
      elsif(slope > 1 && x1 > x2)
        dir = @@N
      elsif(slope > 1 && x1 < x2)
        dir = @@S
      end
    rescue ZeroDivisionError
      if(y1 > y2)
        dir = @@N;
      else
        dir = @@S;
      end
    rescue FloatDomainError
      if(y1 > y2)
        dir = @@N;
      else
        dir = @@S;
      end
    end
    return dir;
  end  

  def get_direction x1,y1,x2,y2
    tile1 = tile_value(x1,y1)
    tile2 = tile_value(x2,y2)
    i = tile1[0]; j = tile1[1]
    i2 = tile2[0]; j2 = tile2[1]
    row_even_directions = [[@@N,i-2,j],[@@S,i+2,j],[@@E,i,j+1],[@@W,i,j-1],[@@NE,i-1,j],[@@NW,i-1,j-1],[@@SE,i+1,j],[@@SW,i+1,j-1]]
    row_odd_directions = [[@@N,i-2,j],[@@S,i+2,j],[@@E,i,j+1],[@@W,i,j-1],[@@NE,i-1,j+1],[@@NW,i-1,j],[@@SE,i+1,j+1],[@@SW,i+1,j]]
    if i%2 == 0  
      0.upto(row_even_directions.length-1) do |k|
        return row_even_directions[k][0] if(i2==row_even_directions[k][1]&&j2==row_even_directions[k][2])
      end
    else
       0.upto(row_odd_directions.length-1) do |k|
        return row_odd_directions[k][0] if(i2==row_odd_directions[k][1]&&j2==row_odd_directions[k][2]) 
      end
    end
    return 0
  end
  
  def get_neighbor i,j,direction
     ret = nil
     if(i%2==0)
      ret = @row_even_directions[direction]
     else 
      ret = @row_odd_directions[direction]
     end
     return [ret[0]+i, ret[1]+j] if(@grid[ret[0]+i] && @grid[ret[0]+i][ret[1]+j]) 
  end
  
  def neighbors node
    i=node.x
    j=node.y
    ret_list = []
    0.upto(7) do |k|
        neighbor = get_neighbor(i,j,k)
        if(!neighbor.nil? && (@grid[neighbor[0]][neighbor[1]].value ==0 || @grid[neighbor[0]][neighbor[1]].target)  && @grid[neighbor[0]][neighbor[1]].terrain_type==0 ) 
          ret_list.push(@grid[neighbor[0]][neighbor[1]])
        end
    end
    ret_list
  end

  def tile_value x,y
    half_x = ((x*2/@tile_width).floor).to_f
    half_y = ((y*2/@tile_height).floor).to_f
    half_w = @tile_width/2
    half_h = @tile_height/2
    up = false
    if(half_x%2==half_y%2)
      if(x-half_x*half_w+y-half_y*half_h < (half_x+1)*half_w-x+(half_y+1)*half_h-y)
        up = true
      end
    else
      if((half_x+1)*half_w-x+y-half_y*half_h <x-half_x*half_w +(half_y+1)*half_h-y)
        up = true
      end
    end
    grid_x=0
    grid_y=0
    if(half_x%2==0&&half_y%2==0&&!up ||half_x%2!=0&&half_y%2==0&&!up || half_x%2!=0&&half_y%2!=0 &&up || half_x%2==0&&half_y%2!=0 &&up)
      grid_y = 2*(half_y/2).floor+1
      grid_x = (half_x/2).floor
    else
      grid_y = half_y + half_y%2
      grid_x = ((half_x+1)/2).floor
    end
    return [grid_y,grid_x]
  end

  def value i,j
    y = (((i-1)*@tile_height/2).round).to_f
    y = 0.0 if(y<0.0)
    x=-((i+1)%2)*@tile_width/2+j*@tile_width+@tile_width/2;
    return [x,y]
  end
  
  def add_object_to_grid obj
    no_of_rows = (obj.xdim/@tile_iso_length).ceil
    no_of_columns = (obj.ydim/@tile_iso_length).ceil
    top_x = obj.owner['coords']['x'] 
    top_y = obj.owner['coords']['y'] -@tile_height/2
    origin_tile = tile_value(top_x,top_y)
    map_tiles = []
    0.upto(no_of_rows) do |i|
      next if(origin_tile.nil?) 
      origin_tile = get_neighbor(origin_tile[0],origin_tile[1],@@SW)
      next if(origin_tile.nil?)
      looping_tile = []
      looping_tile[0] = origin_tile[0];
      looping_tile[1] = origin_tile[1];
      0.upto(no_of_columns) do |j|
        @grid[looping_tile[0]][looping_tile[1]].value = obj
        map_tiles.push(@grid[looping_tile[0]][looping_tile[1]])
        looping_tile = get_neighbor(looping_tile[0],looping_tile[1],@@SE)
      end
    end
    obj.map_tiles = map_tiles
  end

  def validate_location obj
    valid = true
    0.upto(@objects.length) do |i|
      if(obj.collides(@objects[i]))
        valid = false
        break
      end
    end
    inside_polygon = Util.is_inside(obj,@zone_points)
    return valid && inside_polygon
  end
  
  def add_element obj
      @objects.push(obj)
      add_object_to_grid(obj)
      return true
  end

  def lookup_location x,y
    map_coords = tile_value(x,y)
    return @grid[map_coords[0]][map_coords[1]].value.owner if(@grid[map_coords[0]][map_coords[1]].value!=0)
    return nil
  end
  
  def occupied x,y
    map_tiles = tile_value x,y
    return true if(@grid[map_tiles[0]][map_tiles[1]].value!=0)
    return false
  end
  
    
  def move_object object,x,y,options = {}
    ignore_place = false
    callback = nil
    if(options['callback']) 
      callback = options['callback']
    end
    if(options['ignore_place']) 
      ignore_place = options['callback']
    end
    astar = Astar.new
    src_tiles = tile_value(object.coords['x'], object.coords['y'])
    dest_tiles = tile_value(x,y)
    @grid[dest_tiles[0]][dest_tiles[1]].target= true
    if(@grid[dest_tiles[0]][dest_tiles[1]].value!=0 && !options['ignore_place']) 
      dest_tiles = get_nearest_empty_tile(dest_tiles)
    end
    path = astar.get_optimal_path(self,@grid[src_tiles[0]][src_tiles[1]],@grid[dest_tiles[0]][dest_tiles[1]])
    @grid[dest_tiles[0]][dest_tiles[1]].target = false
    0.upto(@grid.length-1) do |i|
      0.upto(@grid[0].length-1) do |j|
          @grid[i][j].g=@grid[i][j].h=@grid[i][j].f = 0
          @grid[i][j].visited=@grid[i][j].closed = false
          @grid[i][j].parent = nil
      end
    end
    if(!path.nil?)
      object.moving = false;
	    object.target_point = {'x'=>x,'y'=>y}
      object.moving_path = path;
      object.movement_finish_callback = callback if(!callback.nil?)
      return path
    end
  end


  def get_nearest_empty_tile tile
    open_list = []
    open_list.push(tile)
    found = false
    while(!found) do
      tile = open_list[0]
      0.upto(7) do |k|
        neighbor = get_neighbor(tile[0],tile[1],k)
        if(!neighbor.nil? && @grid[neighbor[0]][neighbor[1]].value ==0 &&
          @grid[neighbor[0]][neighbor[1]].terrain_type==0)
             return neighbor;
        elsif(!neighbor.nil?) 
          open_list.push(neighbor)
        end
      end
      open_list.shift      
    end 
  end
  
  
  end
end


